src/Xamarin.Android.Build.Tasks/Tasks/BuildApk.cs

```diff
@@ -109,22 +109,58 @@ public class BuildApk : AndroidTask
 
 		protected virtual void FixupArchive (ZipArchiveEx zip) { }
 
+		List<string> existingEntries = new List<string> ();
+
 		void ExecuteWithAbi (string [] supportedAbis, string apkInputPath, string apkOutputPath)
 		{
-			var temp = apkOutputPath + "new";
 			ArchiveFileList files = new ArchiveFileList ();
-			if (apkInputPath != null)
-				File.Copy (apkInputPath, temp, overwrite: true);
+			bool refresh = true;
+			if (apkInputPath != null && File.Exists (apkInputPath) && !File.Exists (apkOutputPath)) {
+				File.Copy (apkInputPath, apkOutputPath, overwrite: true);
+				refresh = false;
+			}
 			using (var notice = Assembly.GetExecutingAssembly ().GetManifestResourceStream ("NOTICE.txt"))
-			using (var apk = new ZipArchiveEx (temp, apkInputPath != null ? FileMode.Open : FileMode.Create )) {
+			using (var apk = new ZipArchiveEx (apkOutputPath, File.Exists (apkOutputPath) ? FileMode.Open : FileMode.Create )) {
+				if (refresh) {
+					for (long i = 0; i < apk.Archive.EntryCount; i++) {
+						ZipEntry e = apk.Archive.ReadEntry ((ulong) i);
+						existingEntries.Add (e.FullName);
+					}
+				}
+				if (apkInputPath != null && File.Exists (apkInputPath) && refresh) {
+					var lastWriteOutput = File.Exists (apkOutputPath) ? File.GetLastWriteTimeUtc (apkOutputPath) : DateTime.MinValue;
+					var lastWriteInput = File.GetLastWriteTimeUtc (apkInputPath);
+					if (lastWriteInput > lastWriteOutput) {
+						using (var packaged = new ZipArchiveEx (apkInputPath, FileMode.Open)) {
+							foreach (var entry in packaged.Archive) {
+								existingEntries.Remove (entry.FullName);
+								if (apk.Archive.ContainsEntry (entry.FullName)) {
+									ZipEntry e = apk.Archive.ReadEntry (entry.FullName);
+									// check the CRC values as the ModifiedDate is always 01/01/1980 in the aapt generated file.
+									if (entry.CRC == e.CRC) {
+										Log.LogDebugMessage ($"Skipping {entry.FullName} from {apkInputPath} as its up to date.");
+										continue;
+									}
+								}
```

Comment:
I think we need a test where we test Add/Remove/Update of `AndroidResource` files to make sure the APK is updated in the right way.

Reply:
Good idea.